File: scripts/build-tree-grid.ts
import fs from "node:fs";
import path from "node:path";

import { TRIGGER_ALIASES } from "@/lib/constants";
import type { TreeGridCell, TreeGridData } from "@/lib/types";

const inputPath = process.argv[2];
const outputPath = process.argv[3] ?? "data/tree-grid.generated.json";

if (!inputPath) {
  console.error("Usage: npm run build-tree-grid -- <street-tree-census.csv> [output.json]");
  process.exit(1);
}

const csv = fs.readFileSync(path.resolve(inputPath), "utf8");
const rows = parseCsv(csv);

const latKey = rows.headers.find((header) => ["latitude", "lat"].includes(header.toLowerCase()));
const lngKey = rows.headers.find((header) => ["longitude", "lng", "lon"].includes(header.toLowerCase()));
const speciesKey = rows.headers.find((header) =>
  ["spc_common", "species", "species_name"].includes(header.toLowerCase()),
);
const areaKey = rows.headers.find((header) => ["nta_name", "boroname", "zip_city"].includes(header.toLowerCase()));

if (!latKey || !lngKey || !speciesKey) {
  console.error("Could not find latitude, longitude, and species columns in the CSV.");
  process.exit(1);
}

const points = rows.records
  .map((record) => ({
    lat: Number(record[latKey]),
    lng: Number(record[lngKey]),
    species: record[speciesKey] ?? "unknown",
    areaName: areaKey ? record[areaKey] ?? "NYC canopy corridor" : "NYC canopy corridor",
  }))
  .filter((record) => !Number.isNaN(record.lat) && !Number.isNaN(record.lng));

if (!points.length) {
  console.error("No valid rows found in the CSV.");
  process.exit(1);
}

const origin = {
  lat: Math.min(...points.map((point) => point.lat)),
  lng: Math.min(...points.map((point) => point.lng)),
};
const latStep = 0.01;
const lngStep = 0.012;

const buckets = new Map<
  string,
  {
    centerLatSum: number;
    centerLngSum: number;
    count: number;
    areaCounts: Map<string, number>;
    speciesCounts: Map<string, number>;
  }
>();

for (const point of points) {
  const latIndex = Math.floor((point.lat - origin.lat) / latStep);
  const lngIndex = Math.floor((point.lng - origin.lng) / lngStep);
  const key = `${latIndex}:${lngIndex}`;
  const bucket = buckets.get(key) ?? {
    centerLatSum: 0,
    centerLngSum: 0,
    count: 0,
    areaCounts: new Map<string, number>(),
    speciesCounts: new Map<string, number>(),
  };

  bucket.centerLatSum += point.lat;
  bucket.centerLngSum += point.lng;
  bucket.count += 1;
  bucket.areaCounts.set(point.areaName, (bucket.areaCounts.get(point.areaName) ?? 0) + 1);
  const trigger = mapSpeciesToTrigger(point.species);
  bucket.speciesCounts.set(trigger, (bucket.speciesCounts.get(trigger) ?? 0) + 1);
  buckets.set(key, bucket);
}

const maxCount = Math.max(...Array.from(buckets.values(), (bucket) => bucket.count));
const cells: TreeGridCell[] = Array.from(buckets.entries()).map(([key, bucket]) => {
  const speciesWeights = Object.fromEntries(
    Array.from(bucket.speciesCounts.entries())
      .sort((a, b) => b[1] - a[1])
      .slice(0, 4)
      .map(([species, count]) => [species, Number((count / bucket.count).toFixed(2))]),
  );

  return {
    key,
    center: {
      lat: Number((bucket.centerLatSum / bucket.count).toFixed(6)),
      lng: Number((bucket.centerLngSum / bucket.count).toFixed(6)),
    },
    areaName: getTopKey(bucket.areaCounts),
    density: Number((bucket.count / maxCount).toFixed(2)),
    canopyScore: Math.round((bucket.count / maxCount) * 80),
    topSpecies: Object.keys(speciesWeights),
    speciesWeights,
  };
});

const treeGrid: TreeGridData = {
  version: new Date().toISOString().slice(0, 10),
  origin,
  latStep,
  lngStep,
  cellSizeMeters: 1100,
  cells,
};

fs.writeFileSync(path.resolve(outputPath), `${JSON.stringify(treeGrid, null, 2)}\n`);
console.log(`Wrote ${cells.length} cells to ${outputPath}`);

function getTopKey(values: Map<string, number>) {
  return Array.from(values.entries()).sort((a, b) => b[1] - a[1])[0]?.[0] ?? "NYC canopy corridor";
}

function mapSpeciesToTrigger(species: string) {
  const normalized = species.toLowerCase();

  for (const [trigger, aliases] of Object.entries(TRIGGER_ALIASES)) {
    if (aliases.some((alias) => normalized.includes(alias))) {
      return trigger;
    }
  }

  return "tree";
}

function parseCsv(input: string) {
  const lines = input.split(/\r?\n/).filter(Boolean);
  const headers = splitCsvLine(lines[0] ?? "");
  const records = lines.slice(1).map((line) => {
    const values = splitCsvLine(line);
    return Object.fromEntries(headers.map((header, index) => [header, values[index] ?? ""]));
  });

  return { headers, records };
}

function splitCsvLine(line: string) {
  const fields: string[] = [];
  let current = "";
  let inQuotes = false;

  for (let index = 0; index < line.length; index += 1) {
    const character = line[index];

    if (character === '"' && line[index + 1] === '"') {
      current += '"';
      index += 1;
      continue;
    }

    if (character === '"') {
      inQuotes = !inQuotes;
      continue;
    }

    if (character === "," && !inQuotes) {
      fields.push(current);
      current = "";
      continue;
    }

    current += character;
  }

  fields.push(current);
  return fields;
}

File: backend/scripts/data/build_tree_grid.py
from __future__ import annotations

import csv
import json
import math
import sys
from collections import defaultdict
from datetime import date
from pathlib import Path

from app.domain.scoring import TRIGGER_ALIASES

REPO_ROOT = Path(__file__).resolve().parents[3]
DEFAULT_OUTPUT_PATH = REPO_ROOT / "data" / "generated" / "tree-grid.generated.json"


def main():
    if len(sys.argv) < 2:
        print("Usage: python backend/scripts/data/build_tree_grid.py <street-tree-census.csv> [output.json]", file=sys.stderr)
        sys.exit(1)

    input_path = Path(sys.argv[1]).resolve()
    output_path = Path(sys.argv[2]).resolve() if len(sys.argv) > 2 else DEFAULT_OUTPUT_PATH

    rows = read_csv(input_path)
    lat_key = find_header(rows["headers"], ["latitude", "lat"])
    lng_key = find_header(rows["headers"], ["longitude", "lng", "lon"])
    species_key = find_header(rows["headers"], ["spc_common", "species", "species_name"])
    area_key = find_header(rows["headers"], ["nta_name", "boroname", "zip_city"])

    if not lat_key or not lng_key or not species_key:
        print("Could not find latitude, longitude, and species columns in the CSV.", file=sys.stderr)
        sys.exit(1)

    points = []
    for record in rows["records"]:
        try:
            lat = float(record.get(lat_key, ""))
            lng = float(record.get(lng_key, ""))
        except ValueError:
            continue

        points.append(
            {
                "lat": lat,
                "lng": lng,
                "species": record.get(species_key, "unknown"),
                "area_name": record.get(area_key, "NYC canopy corridor") if area_key else "NYC canopy corridor",
            }
        )

    if not points:
        print("No valid rows found in the CSV.", file=sys.stderr)
        sys.exit(1)

    origin = {
        "lat": min(point["lat"] for point in points),
        "lng": min(point["lng"] for point in points),
    }
    lat_step = 0.00018
    lng_step = 0.00024
    cell_size_meters = round(((lat_step * 111_000) + (lng_step * 84_000)) / 2)

    buckets: dict[str, dict] = {}
    for point in points:
        lat_index = math.floor((point["lat"] - origin["lat"]) / lat_step)
        lng_index = math.floor((point["lng"] - origin["lng"]) / lng_step)
        key = f"{lat_index}:{lng_index}"
        bucket = buckets.setdefault(
            key,
            {
                "center_lat_sum": 0.0,
                "center_lng_sum": 0.0,
                "count": 0,
                "area_counts": defaultdict(int),
                "species_counts": defaultdict(int),
            },
        )

        bucket["center_lat_sum"] += point["lat"]
        bucket["center_lng_sum"] += point["lng"]
        bucket["count"] += 1
        bucket["area_counts"][point["area_name"]] += 1
        bucket["species_counts"][map_species_to_trigger(point["species"])] += 1

    max_count = max(bucket["count"] for bucket in buckets.values())
    cells = []

    for key, bucket in buckets.items():
        ordered_species = sorted(bucket["species_counts"].items(), key=lambda item: item[1], reverse=True)[:4]
        species_weights = {
            species: round(count / bucket["count"], 2)
            for species, count in ordered_species
        }

        cells.append(
            {
                "key": key,
                "center": {
                    "lat": round(bucket["center_lat_sum"] / bucket["count"], 6),
                    "lng": round(bucket["center_lng_sum"] / bucket["count"], 6),
                },
                "areaName": get_top_key(bucket["area_counts"]),
                "density": round(bucket["count"] / max_count, 2),
                "canopyScore": round((bucket["count"] / max_count) * 80),
                "topSpecies": list(species_weights.keys()),
                "speciesWeights": species_weights,
            }
        )

    tree_grid = {
        "version": date.today().isoformat(),
        "origin": origin,
        "latStep": lat_step,
        "lngStep": lng_step,
        "cellSizeMeters": cell_size_meters,
        "cells": cells,
    }

    output_path.parent.mkdir(parents=True, exist_ok=True)
    output_path.write_text(json.dumps(tree_grid, indent=2) + "\n", encoding="utf-8")
    print(f"Wrote {len(cells)} cells to {output_path}")


def read_csv(input_path: Path):
    with input_path.open("r", encoding="utf-8-sig", newline="") as handle:
        reader = csv.DictReader(handle)
        records = [dict(row) for row in reader]
        return {
            "headers": list(reader.fieldnames or []),
            "records": records,
        }


def find_header(headers: list[str], candidates: list[str]):
    lowered = {header.lower(): header for header in headers}
    for candidate in candidates:
        if candidate in lowered:
            return lowered[candidate]
    return None


def get_top_key(values: dict[str, int]):
    return sorted(values.items(), key=lambda item: item[1], reverse=True)[0][0] if values else "NYC canopy corridor"


def map_species_to_trigger(species: str):
    normalized = species.lower()
    for trigger, aliases in TRIGGER_ALIASES.items():
        if any(alias in normalized for alias in aliases):
            return trigger
    return "tree"


if __name__ == "__main__":
    main()
